File: 2025/java/senior/Senior4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Senior4 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer tokens = new StringTokenizer(reader.readLine());

        int n = Integer.parseInt(tokens.nextToken());
        int m = Integer.parseInt(tokens.nextToken());

        int[][] nodes = new int[m + 1][3];
        List<List<int[]>> graph = new ArrayList<>();
        List<PriorityQueue<int[]>> edges = new ArrayList<>();

        nodes[0] = new int[]{1, 1, 0};
        for (int i = 0; i <= m; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < n; i++) {
            edges.add(new PriorityQueue<>(Comparator.comparingInt(a -> a[1])));
        }
        edges.getFirst().add(new int[]{0, 0});
        for (int i = 1; i <= m; i++) {
            tokens = new StringTokenizer(reader.readLine());
            int s = Integer.parseInt(tokens.nextToken());
            int e = Integer.parseInt(tokens.nextToken());
            int c = Integer.parseInt(tokens.nextToken());

            nodes[i] = new int[]{s, e, c};

            edges.get(s - 1).add(new int[]{i, c});
            edges.get(e - 1).add(new int[]{i, c});
        }

        for (PriorityQueue<int[]> edge : edges) {
            int[] start = edge.poll();
            while (!edge.isEmpty()) {
                int[] end = edge.poll();
                graph.get(start[0]).add(new int[]{end[0], Math.abs(start[1] - end[1])});
                graph.get(end[0]).add(new int[]{start[0], Math.abs(start[1] - end[1])});
                start = end;
            }
        }

        long[] dist = new long[m + 1];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[0] = 0;
        PriorityQueue<long[]> notVisited = new PriorityQueue<>(Comparator.comparingLong(a -> a[0]));
        notVisited.add(new long[]{0, 0});

        while (!notVisited.isEmpty()) {
            long[] next = notVisited.poll();
            for (int[] j : graph.get((int) next[1])) {
                long p = dist[(int) next[1]] + j[1];
                if (p < dist[j[0]]) {
                    dist[j[0]] = p;
                    notVisited.add(new long[]{p, j[0]});
                }
            }
        }

        long min = Long.MAX_VALUE;
        for (int i = 0; i <= m; i++) {
            if (nodes[i][0] == n || nodes[i][1] == n) {
                min = Math.min(min, dist[i]);
            }
        }
        System.out.println(min);
    }
}

File: 2025/cpp/senior/senior4.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n, m;
  cin >> n >> m;

  vector<vector<int>> nodes(m + 1);
  vector<vector<pair<int, int>>> graph(m + 1);
  vector<priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>>> edges(n);

  nodes[0] = {1, 1, 0};
  edges[0].push({0, 0});
  for (int i = 0; i < m; i++) {
    int s, e, c;
    cin >> s >> e >> c;
    nodes[i + 1] = {s, e, c};
    edges[s - 1].push({c, i + 1});
    edges[e - 1].push({c, i + 1});
  }

  for (auto edge : edges) {
    auto start = edge.top();
    edge.pop();
    while (!edge.empty()) {
      auto end = edge.top();
      edge.pop();
      graph[start.second].emplace_back(end.second, abs(start.first - end.first));
      graph[end.second].emplace_back(start.second, abs(end.first - start.first));
      start = end;
    }
  }

  vector<long> dist(m + 1, LONG_MAX);
  dist[0] = 0;
  priority_queue<pair<long, int>, vector<pair<long, int>>, greater<pair<long, int>>> notVisited{};
  notVisited.push({0, 0});

  while (!notVisited.empty()) {
    auto next = notVisited.top();
    notVisited.pop();
    for (auto edge : graph[next.second]) {
      long p = dist[next.second] + edge.second;
      if (p < dist[edge.first]) {
        dist[edge.first] = p;
        notVisited.push({p, edge.first});
      }
    }
  }

  long minDist = LONG_MAX;
  for (int i = 0; i <= m; i++) {
    if (nodes[i][0] == n || nodes[i][1] == n) {
      minDist = min(minDist, dist[i]);
    }
  }
  cout << minDist << endl;
}